File: backend/app/ppt/renderer.py
"""
PPT Renderer — Programmatic PPTX generation from Slide DSL.

Rule: the renderer is NOT an LLM. It follows template constraints.
"""
import asyncio
import io
import logging
import httpx
from pptx import Presentation
from pptx.util import Inches, Pt
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN

from app.core.config import settings

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
if not logger.handlers:
    _h = logging.StreamHandler()
    _h.setFormatter(logging.Formatter('%(levelname)s [ppt] %(message)s'))
    logger.addHandler(_h)

# ---- Image fetching via Tavily ----
_IMAGE_CACHE: dict[str, bytes] = {}
_MAX_IMAGE_CACHE = 64


async def _fetch_images(queries: list[str]) -> list[bytes]:
    """Search images via Tavily and download them in parallel."""
    if not settings.tavily_api_key:
        return await _fallback_images(queries)

    async def _fetch_one(q: str) -> bytes | None:
        if q in _IMAGE_CACHE:
            logger.debug("Image cache hit: %s", q[:60])
            return _IMAGE_CACHE[q]
        try:
            from tavily import TavilyClient
            client = TavilyClient(api_key=settings.tavily_api_key)
            resp = client.search(query=q, search_depth="basic", include_images=True, max_results=3)
            image_urls = resp.get("images", [])
            if image_urls:
                logger.info("Tavily found %d images for query '%s'", len(image_urls), q[:60])
                img_bytes = await _download_image(image_urls[0])
                if img_bytes:
                    _cache_image(q, img_bytes)
                    logger.info("Downloaded image %d bytes from %s", len(img_bytes), image_urls[0][:80])
                    return img_bytes
                else:
                    logger.warning("Failed to download image from %s", image_urls[0][:80])
            else:
                logger.warning("Tavily returned no images for query '%s'", q[:60])
        except Exception:
            logger.exception("Tavily search failed for query '%s'", q[:60])
        return None

    results = await asyncio.gather(*(_fetch_one(q) for q in queries[:3]))
    return [r for r in results if r is not None]


def _cache_image(key: str, data: bytes) -> None:
    if len(_IMAGE_CACHE) >= _MAX_IMAGE_CACHE:
        _IMAGE_CACHE.pop(next(iter(_IMAGE_CACHE)))
    _IMAGE_CACHE[key] = data


async def _download_image(url: str) -> bytes | None:
    try:
        async with httpx.AsyncClient(timeout=15, follow_redirects=True) as client:
            resp = await client.get(url)
            if resp.status_code == 200 and len(resp.content) > 2000:
                return resp.content
    except Exception:
        pass
    return None


async def _fallback_images(queries: list[str]) -> list[bytes]:
    logger.info("Using Unsplash fallback for images (no Tavily API key)")
    async def _fetch_one(q: str) -> bytes | None:
        if q in _IMAGE_CACHE:
            return _IMAGE_CACHE[q]
        try:
            url = f"https://source.unsplash.com/800x600/?{q}"
            async with httpx.AsyncClient(timeout=10, follow_redirects=True) as client:
                resp = await client.get(url)
                if resp.status_code == 200 and len(resp.content) > 2000:
                    _cache_image(q, resp.content)
                    logger.info("Unsplash download success for '%s': %d bytes", q[:60], len(resp.content))
                    return resp.content
                else:
                    logger.warning("Unsplash returned status %d, %d bytes for '%s'",
                                   resp.status_code, len(resp.content), q[:60])
        except Exception:
            logger.exception("Unsplash download failed for '%s'", q[:60])
        return None

    results = await asyncio.gather(*(_fetch_one(q) for q in queries[:2]))
    return [r for r in results if r is not None]


def _embed_image(slide, image_bytes: bytes, left, top, width, height):
    try:
        slide.shapes.add_picture(io.BytesIO(image_bytes),
                                 Inches(left), Inches(top),
                                 Inches(width), Inches(height))
        logger.info("Embedded image at (%.1f, %.1f) size %.1fx%.1f, %d bytes",
                     left, top, width, height, len(image_bytes))
    except Exception as e:
        logger.warning("Failed to embed image (%d bytes), falling back to placeholder: %s",
                       len(image_bytes), e)
        rect = slide.shapes.add_shape(
            1, Inches(left), Inches(top), Inches(width), Inches(height))
        rect.fill.solid()
        rect.fill.fore_color.rgb = _color("E8E8E8")
        rect.line.fill.background()


# ---- Theme System ----
THEMES = {
    "japan": {"bg": "F5F0E8", "accent": "9B1B30", "text": "2D2D2D", "sub": "8B7355"},
    "minimal": {"bg": "FFFFFF", "accent": "111111", "text": "1A1A1A", "sub": "757575"},
    "tropical": {"bg": "FFF8F0", "accent": "E8734A", "text": "2C3E50", "sub": "8E6E53"},
    "ocean": {"bg": "F0F4F8", "accent": "1B4F72", "text": "1C2833", "sub": "5D6D7E"},
    "editorial": {"bg": "FAFAFA", "accent": "2C3E50", "text": "1A1A1A", "sub": "7F8C8D"},
    "nature": {"bg": "F4F6F0", "accent": "3E6B48", "text": "2D3436", "sub": "6B7F6B"},
}

DEFAULT_THEME = THEMES["minimal"]

WIDTH = Inches(13.333)
HEIGHT = Inches(7.5)


def _color(hex_str: str) -> RGBColor:
    return RGBColor(int(hex_str[0:2], 16), int(hex_str[2:4], 16), int(hex_str[4:6], 16))


async def render_pptx(slides: list[dict], theme_name: str = "minimal",
                     assets: list[dict] | None = None) -> bytes:
    prs = Presentation()
    prs.slide_width = WIDTH
    prs.slide_height = HEIGHT

    theme = THEMES.get(theme_name, DEFAULT_THEME)
    assets = assets or []

    blank_layout = prs.slide_layouts[6]

    # Pre-fetch images for ALL slides that have asset queries
    image_futures = {}
    for idx, s in enumerate(slides):
        queries = _get_asset_queries(assets, idx)
        if queries:
            image_futures[idx] = asyncio.ensure_future(_fetch_images(queries))

    for idx, s in enumerate(slides):
        slide_type = s.get("slide_type", "content")
        slide = prs.slides.add_slide(blank_layout)
        _set_bg(slide, theme["bg"])

        images = None
        if idx in image_futures:
            images = await image_futures[idx]

        render = _SLIDE_RENDERERS.get(slide_type, _render_content)
        render(slide, s, theme, images or [])

    # Log image embedding summary
    slides_with_images = [i for i in image_futures]
    logger.info("PPT render done: %d slides, %d with image queries, assets=%s",
                 len(slides), len(slides_with_images),
                 [(a.get("slide_index"), a.get("queries", [])[:1]) for a in assets])

    buf = io.BytesIO()
    prs.save(buf)
    buf.seek(0)
    return buf.read()


def _get_asset_queries(assets: list[dict], slide_index: int) -> list[str]:
    for a in assets:
        if a.get("slide_index") == slide_index:
            return a.get("queries", [])
    return []


_SLIDE_RENDERERS = {}  # populated after all render functions are defined


def _set_bg(slide, color: str):
    background = slide.background
    fill = background.fill
    fill.solid()
    fill.fore_color.rgb = _color(color)


def _add_text_box(slide, left, top, width, height, text, font_size=18, bold=False,
                  color="2D2D2D", align=PP_ALIGN.LEFT, font_name="Arial"):
    txBox = slide.shapes.add_textbox(Inches(left), Inches(top), Inches(width), Inches(height))
    tf = txBox.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    p.text = text
    p.font.size = Pt(font_size)
    p.font.bold = bold
    p.font.color.rgb = _color(color)
    p.font.name = font_name
    p.alignment = align
    return tf


# ---- Slide Renderers ----

def _render_cover(slide, s, theme, images: list[bytes] | None = None):
    images = images or []
    title = s.get("title", "Travel Plan")
    subtitle = s.get("subtitle", "")

    if images:
        _embed_image(slide, images[0], 6.8, 0, 6.6, 7.5)
        panel = slide.shapes.add_shape(1, Inches(0), Inches(0), Inches(7.2), Inches(7.5))
        panel.fill.solid()
        panel.fill.fore_color.rgb = _color(theme["bg"])
        panel.line.fill.background()

    _add_text_box(slide, 0.8, 2.2, 5.5, 1.5, title, 44, True, theme["accent"],
                  PP_ALIGN.LEFT, "Arial Black")
    if subtitle:
        _add_text_box(slide, 0.8, 3.7, 5, 0.8, subtitle, 20, False, theme["sub"])
    line = slide.shapes.add_shape(1, Inches(0.8), Inches(4.7), Inches(3), Inches(0.04))
    line.fill.solid()
    line.fill.fore_color.rgb = _color(theme["accent"])
    line.line.fill.background()


def _render_timeline(slide, s, theme, images: list[bytes] | None = None):
    title = s.get("title", "Itinerary")
    subtitle = s.get("subtitle", "")
    days = s.get("days", [])

    if images:
        _embed_image(slide, images[0], 9.8, 0.3, 3.2, 2.4)

    _add_text_box(slide, 0.8, 0.4, 8.5, 0.7, title, 28, True, theme["text"])
    if subtitle:
        _add_text_box(slide, 0.8, 1.0, 11, 0.5, subtitle, 16, False, theme["sub"])

    n = len(days)
    if n == 0:
        return
    col_w = 11.5 / n
    for i, day in enumerate(days):
        x = 0.8 + i * col_w
        _add_text_box(slide, x, 1.8, col_w - 0.3, 0.4,
                      f"Day {day.get('day', i+1)}", 18, True, theme["accent"])
        _add_text_box(slide, x, 2.2, col_w - 0.3, 0.8,
                      day.get("theme", ""), 14, False, theme["sub"])
        activities = day.get("activities", [])
        y = 3.1
        for act in activities[:4]:
            _add_text_box(slide, x, y, col_w - 0.3, 0.6,
                          f"• {act[:60]}", 11, False, theme["text"])
            y += 0.7


def _render_cards(slide, s, theme, images: list[bytes] | None = None):
    title = s.get("title", "")
    items = s.get("items", [])

    if images:
        _embed_image(slide, images[0], 9.8, 0.3, 3.2, 2.4)

    _add_text_box(slide, 0.8, 0.4, 8.5, 0.7, title, 28, True, theme["text"])

    normalized = []
    for item in items:
        if isinstance(item, str):
            normalized.append({"name": item[:60], "desc": ""})
        elif isinstance(item, dict):
            normalized.append({
                "name": str(item.get("name", item.get("title", "")))[:60],
                "desc": str(item.get("desc", item.get("subtitle", "")))[:120]
            })
    items = normalized

    n = len(items)
    if n == 0:
        return
    cols = 2 if n <= 4 else 3
    col_w = 11.5 / cols
    for i, item in enumerate(items):
        row = i // cols
        col = i % cols
        x = 0.8 + col * col_w
        y = 1.6 + row * 2.5
        card = slide.shapes.add_shape(
            1, Inches(x), Inches(y), Inches(col_w - 0.4), Inches(2.0))
        card.fill.solid()
        card.fill.fore_color.rgb = _color("FFFFFF")
        card.line.color.rgb = _color("E0E0E0")
        card.line.width = Pt(0.5)
        _add_text_box(slide, x + 0.2, y + 0.15, col_w - 0.8, 0.4,
                      item["name"][:40], 16, True, theme["accent"])
        _add_text_box(slide, x + 0.2, y + 0.7, col_w - 0.8, 1.0,
                      item["desc"][:120], 11, False, theme["text"])


def _render_budget(slide, s, theme, images: list[bytes] | None = None):
    title = s.get("title", "Budget")
    items = s.get("items", [])
    total = s.get("total", "")
    _add_text_box(slide, 0.8, 0.4, 8.5, 0.7, title, 28, True, theme["text"])

    headers = ["类别", "明细", "金额"]
    col_widths = [2.5, 6.0, 3.0]
    x_positions = [0.8]
    for w in col_widths[:-1]:
        x_positions.append(x_positions[-1] + w)

    y = 1.6
    for j, h in enumerate(headers):
        _add_text_box(slide, x_positions[j], y, col_widths[j], 0.4,
                      h, 13, True, theme["accent"])
    y += 0.5
    for item in items[:8]:
        if isinstance(item, str):
            _add_text_box(slide, 0.8, y, 11, 0.35, f"• {item[:80]}", 12, False, theme["text"])
        else:
            for j, key in enumerate(["category", "detail", "amount"]):
                _add_text_box(slide, x_positions[j], y, col_widths[j], 0.35,
                              str(item.get(key, ""))[:50], 12, False, theme["text"])
        y += 0.4

    if total:
        _add_text_box(slide, 0.8, y + 0.2, 11, 0.4, total, 16, True, theme["accent"])


def _render_map_slide(slide, s, theme, images: list[bytes] | None = None):
    images = images or []
    title = s.get("title", "Route Map")
    _add_text_box(slide, 0.8, 0.4, 11, 0.7, title, 28, True, theme["text"])

    if images:
        _embed_image(slide, images[0], 0.8, 1.5, 11.5, 5.5)
    else:
        _add_text_box(slide, 0.8, 2.5, 11.5, 4.0,
                      s.get("description", "路线示意图"),
                      14, False, theme["sub"])


def _render_ending(slide, s, theme, images: list[bytes] | None = None):
    _add_text_box(slide, 1.5, 2.8, 10, 1.0, s.get("title", "Bon Voyage! ✈️"),
                  42, True, theme["accent"], PP_ALIGN.CENTER, "Arial Black")
    _add_text_box(slide, 1.5, 3.8, 10, 0.6, s.get("subtitle", "Have a wonderful trip"),
                  20, False, theme["sub"], PP_ALIGN.CENTER)


def _render_content(slide, s, theme, images: list[bytes] | None = None):
    title = s.get("title", "")
    bullets = s.get("bullets", [])
    _add_text_box(slide, 0.8, 0.4, 11, 0.7, title, 28, True, theme["text"])
    y = 1.6
    for b in bullets[:8]:
        _add_text_box(slide, 1.2, y, 10.5, 0.5, f"• {b[:100]}", 14, False, theme["text"])
        y += 0.55


_SLIDE_RENDERERS = {
    "cover": _render_cover,
    "map": _render_map_slide,
    "timeline": _render_timeline,
    "overview": _render_timeline,
    "itinerary": _render_timeline,
    "food": _render_cards,
    "hotel": _render_cards,
    "tips": _render_cards,
    "budget": _render_budget,
    "ending": _render_ending,
}

File: backend/app/ppt/renderer.js
/**
 * PptxGenJS Travel Plan Renderer
 *
 * Reads slide JSON + images from a temp directory, writes .pptx.
 *
 * Usage: node renderer.js <input.json> <output.pptx>
 *
 * JSON schema:
 * {
 *   "theme": {"bg": "FFF8F0", "accent": "E8734A", "text": "2C3E50", "sub": "8E6E53"},
 *   "title": "Trip Title",
 *   "destination": "Phuket",
 *   "duration": "5 days",
 *   "slides": [
 *     {"type": "cover", "title": "...", "subtitle": "...", "image": "/tmp/ppt/slide_0.jpg"},
 *     {"type": "timeline", "title": "...", "subtitle": "...", "days": [...]},
 *     {"type": "cards", "title": "...", "items": [{"name": "...", "desc": "..."}]},
 *     {"type": "budget", "title": "...", "items": [{"category": "...", "detail": "...", "amount": "..."}], "total": "..."},
 *     {"type": "map", "title": "...", "image": "/tmp/ppt/slide_N.jpg"},
 *     {"type": "ending", "title": "...", "subtitle": "..."}
 *   ]
 * }
 */

const pptxgen = require("pptxgenjs");
const fs = require("fs");

const [inputPath, outputPath] = process.argv.slice(2);
if (!inputPath || !outputPath) {
  console.error("Usage: node renderer.js <input.json> <output.pptx>");
  process.exit(1);
}

const data = JSON.parse(fs.readFileSync(inputPath, "utf-8"));
const theme = data.theme || {};
const slides = data.slides || [];

const BG = theme.bg || "FFFFFF";
const ACCENT = theme.accent || "111111";
const TEXT = theme.text || "1A1A1A";
const SUB = theme.sub || "757575";
const LIGHT = "F5F5F5";

const pres = new pptxgen();
pres.layout = "LAYOUT_WIDE"; // 13.3" x 7.5"
pres.author = "Travel Advisor";
pres.title = data.title || "Travel Plan";

// ---- Helpers ----

function addImageSafe(slide, path, opts) {
  if (path && fs.existsSync(path)) {
    try {
      slide.addImage({ path, ...opts });
      return true;
    } catch (e) {
      console.error(`Failed to add image ${path}: ${e.message}`);
    }
  }
  return false;
}

// ---- Slide Renderers ----

function renderCover(slide, s) {
  slide.background = { color: BG };

  if (s.image) {
    // Full right-half image
    addImageSafe(slide, s.image, { x: 6.65, y: 0, w: 6.65, h: 7.5, sizing: { type: "cover", w: 6.65, h: 7.5 } });
    // Left panel overlay for text
    slide.addShape(pres.shapes.RECTANGLE, {
      x: 0, y: 0, w: 7.1, h: 7.5, fill: { color: BG },
    });
  }

  slide.addText(s.title || "Travel Plan", {
    x: 0.8, y: 2.0, w: 5.5, h: 1.5,
    fontSize: 44, fontFace: "Arial Black", bold: true,
    color: ACCENT, align: "left", margin: 0,
  });

  if (s.subtitle) {
    slide.addText(s.subtitle, {
      x: 0.8, y: 3.6, w: 5.0, h: 0.8,
      fontSize: 20, fontFace: "Arial", color: SUB, margin: 0,
    });
  }

  // Accent line
  slide.addShape(pres.shapes.RECTANGLE, {
    x: 0.8, y: 4.6, w: 2.5, h: 0.04, fill: { color: ACCENT },
  });
}

function renderTimeline(slide, s) {
  slide.background = { color: BG };

  slide.addText(s.title || "Itinerary", {
    x: 0.8, y: 0.4, w: 11.5, h: 0.7,
    fontSize: 30, fontFace: "Arial Black", bold: true, color: TEXT, margin: 0,
  });
  if (s.subtitle) {
    slide.addText(s.subtitle, {
      x: 0.8, y: 1.05, w: 11.5, h: 0.5,
      fontSize: 16, fontFace: "Arial", color: SUB, margin: 0,
    });
  }

  const days = s.days || [];
  if (days.length === 0) return;

  const colW = 11.5 / days.length;
  days.forEach((day, i) => {
    const x = 0.8 + i * colW;

    // Day number circle
    slide.addShape(pres.shapes.OVAL, {
      x: x, y: 1.8, w: 0.45, h: 0.45, fill: { color: ACCENT },
    });
    slide.addText(String(day.day || i + 1), {
      x: x, y: 1.8, w: 0.45, h: 0.45,
      fontSize: 14, fontFace: "Arial", bold: true, color: "FFFFFF",
      align: "center", valign: "middle", margin: 0,
    });

    slide.addText(day.theme || "", {
      x: x, y: 2.4, w: colW - 0.3, h: 0.8,
      fontSize: 14, fontFace: "Arial", bold: true, color: ACCENT, margin: 0,
    });

    const activities = day.activities || [];
    let y = 3.3;
    activities.slice(0, 5).forEach((act) => {
      slide.addText(String(act).substring(0, 70), {
        x: x, y: y, w: colW - 0.3, h: 0.55,
        fontSize: 11, fontFace: "Arial", color: TEXT,
        bullet: true, margin: 0,
      });
      y += 0.6;
    });
  });
}

function renderCards(slide, s) {
  slide.background = { color: BG };

  slide.addText(s.title || "", {
    x: 0.8, y: 0.4, w: 11.5, h: 0.7,
    fontSize: 30, fontFace: "Arial Black", bold: true, color: TEXT, margin: 0,
  });

  const items = (s.items || []).map((item) => {
    if (typeof item === "string") return { name: item.substring(0, 60), desc: "" };
    return {
      name: String(item.name || item.title || "").substring(0, 50),
      desc: String(item.desc || item.subtitle || "").substring(0, 120),
    };
  });

  if (items.length === 0) return;

  const cols = items.length <= 4 ? 2 : 3;
  const colW = 11.5 / cols;
  const rows = Math.ceil(items.length / cols);

  items.forEach((item, i) => {
    const row = Math.floor(i / cols);
    const col = i % cols;
    const x = 0.8 + col * colW;
    const y = 1.5 + row * 2.5;
    const cardW = colW - 0.4;

    // Card background
    slide.addShape(pres.shapes.RECTANGLE, {
      x: x, y: y, w: cardW, h: 2.1,
      fill: { color: "FFFFFF" },
      shadow: { type: "outer", blur: 6, offset: 2, angle: 135, color: "000000", opacity: 0.08 },
    });

    slide.addText(item.name, {
      x: x + 0.2, y: y + 0.15, w: cardW - 0.4, h: 0.45,
      fontSize: 16, fontFace: "Arial", bold: true, color: ACCENT, margin: 0,
    });
    if (item.desc) {
      slide.addText(item.desc, {
        x: x + 0.2, y: y + 0.7, w: cardW - 0.4, h: 1.2,
        fontSize: 11, fontFace: "Arial", color: TEXT, margin: 0,
      });
    }
  });
}

function renderBudget(slide, s) {
  slide.background = { color: BG };

  slide.addText(s.title || "Budget", {
    x: 0.8, y: 0.4, w: 11.5, h: 0.7,
    fontSize: 30, fontFace: "Arial Black", bold: true, color: TEXT, margin: 0,
  });

  const headers = ["Category", "Details", "Amount"];
  const colWidths = [2.5, 6.0, 3.0];
  const xPositions = [0.8];
  colWidths.slice(0, -1).forEach((w) => xPositions.push(xPositions[xPositions.length - 1] + w));

  const headerRow = headers.map((h, j) => ({
    text: h,
    options: { fontSize: 13, fontFace: "Arial", bold: true, color: ACCENT },
  }));

  const tableData = [headerRow];

  (s.items || []).slice(0, 10).forEach((item) => {
    if (typeof item === "string") {
      tableData.push([{ text: item.substring(0, 80), options: { colspan: 3, fontSize: 12, fontFace: "Arial", color: TEXT } }]);
    } else {
      tableData.push([
        { text: String(item.category || "").substring(0, 30), options: { fontSize: 12, fontFace: "Arial", color: TEXT } },
        { text: String(item.detail || "").substring(0, 50), options: { fontSize: 12, fontFace: "Arial", color: TEXT } },
        { text: String(item.amount || "").substring(0, 30), options: { fontSize: 12, fontFace: "Arial", color: TEXT } },
      ]);
    }
  });

  slide.addTable(tableData, {
    x: 0.8, y: 1.5, w: 11.5,
    colW: colWidths,
    border: { pt: 0.5, color: "E0E0E0" },
    rowH: [0.4],
    autoPage: false,
  });

  // Total row highlight
  if (s.total) {
    slide.addShape(pres.shapes.RECTANGLE, {
      x: 0.8, y: 1.5 + 0.4 * (tableData.length + 0.5), w: 11.5, h: 0.45,
      fill: { color: LIGHT },
    });
    slide.addText(s.total, {
      x: 0.8, y: 1.5 + 0.4 * (tableData.length + 0.5), w: 11.5, h: 0.45,
      fontSize: 16, fontFace: "Arial", bold: true, color: ACCENT,
      align: "center", valign: "middle", margin: 0,
    });
  }
}

function renderMap(slide, s) {
  slide.background = { color: BG };

  slide.addText(s.title || "Route Map", {
    x: 0.8, y: 0.4, w: 11.5, h: 0.7,
    fontSize: 30, fontFace: "Arial Black", bold: true, color: TEXT, margin: 0,
  });

  if (s.image) {
    const added = addImageSafe(slide, s.image, {
      x: 1.5, y: 1.5, w: 10.3, h: 5.2,
      sizing: { type: "contain", w: 10.3, h: 5.2 },
    });
    if (!added) {
      slide.addText(s.description || "Route map image", {
        x: 1.5, y: 3.0, w: 10.3, h: 2.0,
        fontSize: 14, fontFace: "Arial", color: SUB, align: "center", margin: 0,
      });
    }
  }
}

function renderEnding(slide, s) {
  slide.background = { color: ACCENT };

  slide.addText(s.title || "Bon Voyage! ✈️", {
    x: 1.5, y: 2.5, w: 10.3, h: 1.2,
    fontSize: 42, fontFace: "Arial Black", bold: true, color: "FFFFFF",
    align: "center", valign: "middle", margin: 0,
  });
  slide.addText(s.subtitle || "Have a wonderful trip", {
    x: 1.5, y: 3.8, w: 10.3, h: 0.7,
    fontSize: 20, fontFace: "Arial", color: "FFFFFF",
    align: "center", valign: "middle", margin: 0,
  });
}

function renderContent(slide, s) {
  slide.background = { color: BG };

  slide.addText(s.title || "", {
    x: 0.8, y: 0.4, w: 11.5, h: 0.7,
    fontSize: 30, fontFace: "Arial Black", bold: true, color: TEXT, margin: 0,
  });

  const bullets = s.bullets || [];
  const textItems = bullets.slice(0, 10).map((b, i) => ({
    text: String(b).substring(0, 120),
    options: { bullet: true, breakLine: i < bullets.length - 1, fontSize: 14, fontFace: "Arial", color: TEXT },
  }));

  slide.addText(textItems, {
    x: 1.2, y: 1.5, w: 10.5, h: 5.5,
    valign: "top", margin: 0,
  });
}

// ---- Render Dispatch ----

const RENDERERS = {
  cover: renderCover,
  timeline: renderTimeline,
  overview: renderTimeline,
  itinerary: renderTimeline,
  food: renderCards,
  hotel: renderCards,
  tips: renderCards,
  budget: renderBudget,
  map: renderMap,
  ending: renderEnding,
  content: renderContent,
};

// ---- Build Presentation ----

slides.forEach((s) => {
  const slide = pres.addSlide();
  const render = RENDERERS[s.type] || renderContent;
  render(slide, s);
});

// ---- Write Output ----

pres.writeFile({ fileName: outputPath }).then(() => {
  console.log(`PPTX written: ${outputPath}`);
}).catch((err) => {
  console.error("PPTX generation failed:", err);
  process.exit(1);
});
